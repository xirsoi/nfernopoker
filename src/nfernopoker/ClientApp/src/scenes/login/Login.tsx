import * as React from "react";
import { Button, TextField, CardContent, CardActions, Typography } from '@material-ui/core';
import { Component, MouseEvent, ChangeEvent } from "react";
import { compose } from "redux";
import { connect } from "react-redux";
import { RouteComponentProps, withRouter } from "react-router";
import { withFirebase } from "react-redux-firebase";
import { sendMessageAction } from "./actions";

interface ILoginProps {
  classes: any;
  secondaryButtonText: string;
  onSecondaryButton: () => void;
}

interface ISnackMessageProps {
  sendMessage: (message: string) => void;
}

interface ILoginState {
  email: string;
  password: string;
}
interface IFirebase {
  firebase: any;
}
type IProps = ILoginProps & ISnackMessageProps & IFirebase & RouteComponentProps<any>;

class LoginComponent extends Component<IProps, ILoginState> {

  public state: ILoginState

  constructor(
    public props: IProps
  ) {
    super(props);
    this.state = {
      email: "",
      password: ""
    };
  }

  public handleFormChanged(event: ChangeEvent<HTMLInputElement>, name: string): void {
    let newState = { ...this.state };
    newState[name] = event.target.value;
    this.setState(newState);
  }

  public login = (event: MouseEvent<any>) => {
    event.preventDefault();
    event.stopPropagation();

    this.props.firebase.login({
      email: this.state.email,
      password: this.state.password
    }).then(() => {
      this.props.history.push('/games');
    }).catch((ex: any) => {
      this.props.sendMessage(ex.message);
    });
  }

  public render() {
    const { classes } = this.props;
    let isDisabled = !(this.state.email && this.state.password);
    return (
      <form id="login-form" onSubmit={this.login}>
        <CardContent>
          <Typography className={classes.title} color="textSecondary">
            Log in to feel the burn
          </Typography>
          <TextField
            id="email"
            autoComplete="current-user"
            fullWidth={true}
            label="Email"
            onChange={(event: ChangeEvent<HTMLInputElement>) => this.handleFormChanged(event, "email")}
            className={classes.button}
          />
          <TextField
            id="password"
            fullWidth={true}
            type="password"
            label="Password"
            onChange={(event: ChangeEvent<HTMLInputElement>) => this.handleFormChanged(event, "password")}
            className={classes.button}
          />
        </CardContent>
        <CardActions>
          <Button
            id="submit"
            type="submit"
            disabled={isDisabled}
            className={classes.button}
            variant="contained"
            style={{ marginLeft: '16px' }} title="Login" color="primary">Login</Button>
          <Button
            id="secondary-btn"
            onClick={this.props.onSecondaryButton}
            size="small">
            {this.props.secondaryButtonText}
          </Button>
        </CardActions>
      </form>
    );
  }
}

export const LoginTestComp = LoginComponent;

export const Login: React.ComponentClass<ILoginProps> = compose<React.ComponentClass<ILoginProps>>(
  withFirebase,
  withRouter,
  connect(null, { sendMessage: sendMessageAction })
)(LoginComponent)
