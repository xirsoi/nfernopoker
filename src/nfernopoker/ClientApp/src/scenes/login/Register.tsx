import * as React from "react";
import { Component, ChangeEvent } from 'react';
import { compose } from "redux";
import { connect } from "react-redux";
import { withFirebase } from "react-redux-firebase";
import { withRouter, RouteComponentProps } from "react-router";
import { Button, TextField, CardContent, CardActions, Typography } from "@material-ui/core";
import { sendMessageAction } from "./actions";

interface ILocalProps {
  classes: any;
  secondaryButtonText: string;
  onSecondaryButton: () => void;
}
interface ILocalState {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
}
interface IFirebase {
  firebase: any;
}
interface IConnectedDispatch {
  sendMessage: (message: string) => void
}
type IProps = ILocalProps & IConnectedDispatch & RouteComponentProps<any> & IFirebase;

class RegisterComponent extends Component<IProps, ILocalState> {

  constructor(
    public props: IProps
  ) {
    super(props);

    this.state = {
      firstName: '',
      lastName: '',
      email: '',
      password: ''
    };
  }

  public handleFormChanged(event: ChangeEvent<HTMLInputElement>, name: string): void {
    let newState = { ...this.state };
    newState[name] = event.target.value;
    this.setState(newState);
  }

  public handleSubmit = (event: ChangeEvent<HTMLInputElement>) => {
    event.preventDefault();
    event.stopPropagation();
    this.props.firebase.createUser(
      { email: this.state.email, password: this.state.password },
      { firstName: this.state.firstName, lastName: this.state.lastName, email: this.state.email }
    ).then((r: any) => {
      this.props.history.push("/games");
    }).catch((ex: any) => this.props.sendMessage(ex.message));
  }

  public render() {
    let isDisabled = !(this.state.firstName && this.state.lastName && this.state.email && this.state.password);
    return (
      <React.Fragment>
        <form id="register-form" onSubmit={(event: any) => this.handleSubmit(event)}>
          <CardContent>
            <Typography className={this.props.classes.title} color="textSecondary">
              Create a new account
          </Typography>
            <TextField
              id="first-name"
              required={true}
              fullWidth={true}
              label="First Name"
              onChange={(event: ChangeEvent<HTMLInputElement>) => this.handleFormChanged(event, "firstName")}
              className={this.props.classes.button}
            />
            <TextField
              id="last-name"
              required={true}
              fullWidth={true}
              label="Last Name"
              onChange={(event: ChangeEvent<HTMLInputElement>) => this.handleFormChanged(event, "lastName")}
              className={this.props.classes.button}
            />
            <TextField
              id="email"
              required={true}
              fullWidth={true}
              type="email"
              label="Email"
              onChange={(event: ChangeEvent<HTMLInputElement>) => this.handleFormChanged(event, "email")}
              className={this.props.classes.button}
            />
            <TextField
              id="password"
              required={true}
              fullWidth={true}
              type="password"
              label="Password"
              onChange={(event: ChangeEvent<HTMLInputElement>) => this.handleFormChanged(event, "password")}
              className={this.props.classes.button}
            />
          </CardContent>
          <CardActions>
            <Button
              id="submit"
              form="register-form"
              type="submit"
              className={this.props.classes.button}
              disabled={isDisabled}
              variant="contained"
              style={{ marginLeft: '16px' }}
              title="Register"
              color="primary">
              Register
            </Button>
            <Button
              id="secondary-btn"
              onClick={this.props.onSecondaryButton}
              size="small">
              {this.props.secondaryButtonText}
            </Button>
          </CardActions>
        </form>
      </React.Fragment>
    );
  }
}

export const RegisterTestComp = RegisterComponent;

export const Register: React.ComponentClass<ILocalProps> = compose<React.ComponentClass<ILocalProps>>(
  connect(null, { sendMessage: sendMessageAction }),
  withRouter,
  withFirebase
)(RegisterComponent)
