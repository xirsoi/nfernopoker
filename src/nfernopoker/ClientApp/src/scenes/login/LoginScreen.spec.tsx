import * as React from "react";
import { shallow, ShallowWrapper } from "enzyme";

import * as enzyme from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import { LoginScreenTestComp } from "./LoginScreen";
import { Login } from "./Login";
import { Register } from "./Register";
import { Grid } from "@material-ui/core";

enzyme.configure({ adapter: new Adapter() })

let component: ShallowWrapper<any, any, any>;

const fakeProps = {
  classes: {}
};

describe("Login Screen", () => {

  beforeEach(() => {
    component = shallow(<LoginScreenTestComp {...fakeProps} />);
  });

  it("should render without error", () => {
    expect(component.instance()).toMatchSnapshot();
  });

  it("should always render grid content", () => {
    const divs = component.find(Grid);
    expect(divs.length).toBeGreaterThan(0);
  });

  describe("and isLogin", () => {

    beforeEach(() => {
      component.setState({ isLogin: true });
    });

    it("should display the login componet", () => {
      expect(component.find(Login).exists()).toBeTruthy();
      expect(component.find(Register).exists()).toBeFalsy();
    });

  });

  describe("and isLogin is false", () => {

    beforeEach(() => {
      component.setState({ isLogin: false });
    });

    it("should display the register componet", () => {
      expect(component.find(Login).exists()).toBeFalsy();
      expect(component.find(Register).exists()).toBeTruthy();
    });

  });

});
