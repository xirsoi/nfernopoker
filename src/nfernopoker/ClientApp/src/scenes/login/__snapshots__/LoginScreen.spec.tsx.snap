// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Screen should render without error 1`] = `
LoginScreen {
  "context": Object {},
  "props": Object {
    "classes": Object {},
  },
  "refs": Object {},
  "setState": [Function],
  "state": Object {
    "isLogin": true,
  },
  "toggleState": [Function],
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_element": <LoginScreen
        classes={Object {}}
      />,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_newState": null,
      "_rendered": <WithStyles(Grid)
        alignContent="center"
        alignItems="center"
        container={true}
        direction="row"
        justify="center"
      >
        <WithStyles(Grid)
          item={true}
        >
          <WithStyles(Card)>
            <withFirebase(withRouter(Connect(LoginComponent)))
              classes={Object {}}
              onSecondaryButton={[Function]}
              secondaryButtonText="No account?"
            />
          </WithStyles(Card)>
        </WithStyles(Grid)>
      </WithStyles(Grid)>,
      "_rendering": false,
      "_updater": [Circular],
    },
  },
  Symbol(enzyme.__setState__): [Function],
}
`;
